### PR TITLE
[`FA-2`] Add Flash Attention to `Phi`

### PR DESCRIPTION
**Pull request by susnato**
# What does this PR do?

This PR adds Flash Attention to `Phi`.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

cc: @younesbelkada, @ArthurZucker

### CODE REVIEW

**Comment by susnato** [comment]
All the FA tests pass except the `test_flash_attn_2_generate_padding_right`.

Screenshot from 2023-11-23 00-31-04

This is odd given that the `flash_attn_2_inference_padding_right` test is passing as does the `test_flash_attn_2_generate_left_padding` test.

**Comment by younesbelkada** [comment]
@susnato can you try to run that test multiple times? sometimes it is flaky - apart from that the changes look great on my end !

**Comment by susnato** [comment]
Hi @younesbelkada, I ran that test 30 times and every time it failed!

Shouldn't the inference test fail too, if the generation test fails? :sweat_smile: But that does not seem to be the case...

**Comment by younesbelkada** [comment]
Hmm yes correct, what I did for llama was to overwrite the test as can be see here: https://github.com/huggingface/transformers/blob/main/tests/models/llama/test_modeling_llama.py#L392 using a real checkpoint. It would be great if you can do the same and test the next 10 tokens are the same (make sure to use `do_sample=False`)

**Comment by susnato** [comment]
Hi @younesbelkada, thanks a lot for the advice! All the flash attention tests are passing now. :hugs: 

 
Screenshot from 2023-11-23 17-57-49

File: src/transformers/models/phi/modeling_phi.py [hunk 1]
```diff
@@ -361,10 +381,218 @@ def forward(
...
+        key_states = key_states.transpose(1, 2).view(bsz, tgt_len, self.num_heads, self.head_dim)
+        value_states = value_states.transpose(1, 2).view(bsz, tgt_len, self.num_heads, self.head_dim)
+
+        attn_dropout = self.config.attention_dropout if self.training else 0.0
+
+        # In PEFT, usually we cast the layer norms in float32 for training stability reasons
+        # therefore the input hidden states gets silently casted in float32. Hence, we need
+        # cast them back in the correct dtype just to be sure everything works as expected.
+        # This might slowdown training & inference so it is recommended to not cast the LayerNorms
+        # in fp32. (LlamaRMSNorm handles it correctly)
```
Comment (younesbelkada) [inline src/transformers/models/phi/modeling_phi.py:463]:
```suggestion
        # in fp32. (PhiRMSNorm handles it correctly)
```

**Review by younesbelkada — APPROVED** [review]
Truly amazing work @susnato ! Thanks a lot for this great contribution

File: src/transformers/models/phi/modeling_phi.py (hunk 1 above)
Comment (susnato) [reply src/transformers/models/phi/modeling_phi.py:463]: Hey, we don't have a `PhiRMSNorm`, only nn.LayerNorm in the Attention Layer. So removing this part of the line (~PhiRMSNorm handles it correctly~).

Comment (younesbelkada) [reply src/transformers/models/phi/modeling_phi.py:463]: ok makes sense!

File: docs/source/en/model_doc/phi.md [hunk 2]
```diff
@@ -94,6 +94,40 @@ The original code for Phi-1 and Phi-1.5 can be found [here](https://huggingface.
...
+
+>>> # apply the tokenizer.
+>>> tokens = tokenizer(prompt, return_tensors="pt").to("cuda")
+
+>>> # use the model to generate new tokens.
+>>> generated_output = model.generate(**tokens, use_cache=True, max_new_tokens=10)
+
+>>> tokenizer.batch_decode(generated_output)[0]
+'If I were an AI that had just achieved a breakthrough in machine learning, I would be thrilled'
+```
```
Comment (ArthurZucker) [inline docs/source/en/model_doc/phi.md:128]: As we are trying to do with other models, let's add the expected performances gains with tested architecture

File: src/transformers/models/phi/modeling_phi.py [hunk 3]
```diff
@@ -192,7 +212,6 @@ def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
         return hidden_states
 
 
-# Copied from transformers.models.persimmon.modeling_persimmon.PersimmonAttention with Persimmon->Phi,persimmon->phi
```
Comment (ArthurZucker) [inline src/transformers/models/phi/modeling_phi.py:195]: that's not something we want to remove, we should also have flash attention support in persimmon should be the same as this

**Review by ArthurZucker — COMMENTED** [review]
LGTM other than the small comments

File: docs/source/en/model_doc/phi.md (hunk 2 above)
Comment (susnato) [reply docs/source/en/model_doc/phi.md:128]: Can you please check the expected gains for this model too @younesbelkada?

File: src/transformers/models/phi/modeling_phi.py (hunk 3 above)
Comment (susnato) [reply src/transformers/models/phi/modeling_phi.py:195]: Actually, there is another [PR](https://github.com/huggingface/transformers/pull/27052) which is adding the FA support for Persimmon. 

Should I add the `self.causal=True` in the `PersimmonAttention` so that we can keep this `# Copied from` statement?

Comment (ArthurZucker) [reply src/transformers/models/phi/modeling_phi.py:195]: Yeah! 🤗

Comment (susnato) [reply src/transformers/models/phi/modeling_phi.py]: Done @ArthurZucker.

**Review by younesbelkada — APPROVED** [review]
Thanks ! We should be able to merge after accepting the suggestion below

**Comment by susnato** [comment]
Hi @younesbelkada, I have pushed the suggestion you asked.

**Comment by susnato** [comment]
BTW when is the next release date for `transformers`?

**Review by younesbelkada — COMMENTED** [review]
Hi @susnato sorry one last thing before merging, can you apply changes similar to this commit: https://github.com/huggingface/transformers/pull/26463/commits/93fe356f85d05ea275f7be532b9938d4664014f5 and merge with upstream main branch?

**Comment by susnato** [comment]
Just force-pushed the branch along with the changes. @younesbelkada

File: tests/models/phi/test_modeling_phi.py [hunk 4]
```diff
@@ -351,6 +361,43 @@ def test_phi_sequence_classification_model_for_multi_label(self):
         result = model(input_ids, attention_mask=attention_mask, labels=sequence_labels)
         self.assertEqual(result.logits.shape, (self.model_tester.batch_size, self.model_tester.num_labels))
 
+    @require_flash_attn
+    @require_torch_gpu
+    @require_bitsandbytes
+    @pytest.mark.flash_attn_test
+    @slow
+    # Copied from tests.models.llama.test_modeling_llama.LlamaModelTest.test_flash_attn_2_generate_padding_right with LlamaForCausalLM->PhiForCausalLM,LlamaTokenizer->AutoTokenizer,meta-llama/Llama-2-7b-hf->susnato/phi-1_5_dev
+    def test_flash_attn_2_generate_padding_right(self):
```
Comment (younesbelkada) [inline tests/models/phi/test_modeling_phi.py:370]: Nice!

**Review by younesbelkada — APPROVED** [review]
Thanks ! 
We'll probably be able to make it for the next release wdyt @ArthurZucker

**Review by ArthurZucker — APPROVED** [review]
LGTM cc @fxmarty for your sdpa PR that will need rebasing I think

**Comment by fxmarty** [comment]
yep
